### DISCUSSION TITLE
Problematic update

### DISCUSSION THREAD

**Discussion by petar-dambovaliev** [rwf2/Rocket#1991] [Questions]
Hello,

I have an application that was locked onto a git commit of rocket.
I am updating it to the latest release on crates.io
I had to refactor a whole lotta of code.
Currently, i have a problem with the database fairing, whatever that means.
I assume that is dependency injection.
I get the error "missing database fairing for `MyDB`".
This shouldn't happen because i am actually adding it to the app, as you can see from the code below.

```
#[database("datatypes")]
pub struct MyDB(Client);

    let rocket = rocket::custom(config)
        .attach(MyDB::fairing())

//here is the error
MyDB::get_one(&rocket)
```

I also can see this when the application is starting.

```
📡 Fairings:
   >> 'datatypes' Database Pool (ignite)
```

**Comment by p-fruck** [comment]
You are referring to rocket `v0.5-rc.1` if I understood correctly. Do you have a `Rocket.toml` in your project? As described  [here](https://rocket.rs/master/guide/state/#usage), you should create a Rocket.toml with the following content:
```toml
[global.databases]
datatypes={url="mysql://<user>:<password>@<host>/<database>"}
```
Note that this is an example for mysql only, for different types of databases you might have to adapt the url string. You can also define the database url as a environment variable, e.g. 
```env
ROCKET_DATABASES={datatypes={url="mysql://<user>:<password>@<host>/<database>"}}
```

**Reply by petar-dambovaliev** [reply]
I do have this in the `Rocket.toml`
```
[global.databases]
datatypes = {url = "postgres://postgres:postgres@localhost/name", pool_size = 5}
```
If the object is missing because it couldn't be constructed, i would assume that it would give me another error.

I noticed that the application is working.
Only the tests which use the rocket instance are failing in this way.

**Comment by ELD** [comment]
Could you post up:
1. The function where you start up the Rocket instance (with anything sensitive redacted) and the imports in that file
2. The actual compiler error you're receiving

~From first glance, it looks like you don't have a `#[derive(Database)]` on the `MyDB` type. I can't be certain, though, because I don't have enough context about what your code currently looks like.~

**Reply by petar-dambovaliev** [reply]
All tests, like this one, that use to fetch objects from the rocket instance cannot find them.
The actual application works.
```
#[rocket::async_test]
    async fn list_datatypes() -> Result<()> {
        let r = webserver(None);
        let typesdb = DataTypesDb::get_one(&r).await.unwrap();
     }
```

Here is the db definition
```
use rocket_sync_db_pools::{database, postgres::Client};

#[database("datatypes")]
pub struct DataTypesDb(Client);
```

Here is the function that creates the rocket instance
```
pub fn webserver(config: Option<rocket::figment::Figment>) -> rocket::Rocket<rocket::Build> {
    let config = config.unwrap_or_else(rocket::Config::figment);

    let rocket = rocket::custom(config)
        .attach(DataTypesDb::fairing()).....
        }
```

**Reply by ELD** [reply]
Thanks, this helps. Could you copy-paste the compiler error you’re receiving, too?

**Reply by petar-dambovaliev** [reply]
The line with the error is 
```
let typesdb = DataTypesDb::get_one(&r).await.unwrap();
```
The error message
```
---- api::control::tests::list_datatypes stdout ----
   >> missing database fairing for `db::DataTypesDb`
thread 'api::control::tests::list_datatypes' panicked at 'called `Option::unwrap()` on a `None` value'
```

**Reply by ELD** [reply]
Ah, I see now.

So, the issue appears to be that you're returning a `Rocket<Build>`. As per the docs, fairings don't attach until `Rocket<Ignite>` (https://api.rocket.rs/master/rocket/struct.Rocket.html#method.ignite). If you look at the traits for connection pools, you'll find that the fairing returned are `on_ignite` fairings (https://github.com/SergioBenitez/Rocket/blob/master/contrib/sync_db_pools/lib/src/connection.rs#L71-L94), so you have to call `ignite` on the `Rocket<Build>` in order to force fairing attachment.

So, if you change your test code to do:
```rust
...
let r = webserver(None).ignite().await.unwrap();
let typesdb = DataTypesDb::get_one(&r).await.unwrap();
...
```

... You should be good to go.

**Reply by petar-dambovaliev** [reply]
Ohh, thanks for the tip.
It breaks our code a lot though.
It took me thousands of changes (1 week) to arrive to this point.
And now, i need to redesign all of our tests.
It's really unfortunate.

**Reply by ELD** [reply]
It sounds like the commit you were on before your updated was prior to introducing type states. That's going to change a fair amount of code since `Build`, `Ignite`, and `Orbit` mean different things. The [Rocket](https://api.rocket.rs/master/rocket/struct.Rocket.html) type docs should help with the type state changes.

**Reply by petar-dambovaliev** [reply]
https://github.com/SergioBenitez/Rocket?branch=master#8d4d01106e2e10b08100805d40bfa19a7357e900
 We were on this commit

**Reply by petar-dambovaliev** [reply]
> Ah, I see now.
> 
> So, the issue appears to be that you're returning a `Rocket<Build>`. As per the docs, fairings don't attach until `Rocket<Ignite>` (https://api.rocket.rs/master/rocket/struct.Rocket.html#method.ignite). If you look at the traits for connection pools, you'll find that the fairing returned are `on_ignite` fairings (https://github.com/SergioBenitez/Rocket/blob/master/contrib/sync_db_pools/lib/src/connection.rs#L71-L94), so you have to call `ignite` on the `Rocket<Build>` in order to force fairing attachment.
> 
> So, if you change your test code to do:
> 
> ```rust
> ...
> let r = webserver(None).ignite().await.unwrap();
> let typesdb = DataTypesDb::get_one(&r).await.unwrap();
> ...
> ```
> 
> ... You should be good to go.

Why does the library macro now then not implement the type requirement of `Rocket<Ignite>` or a trait that allows for fairings ?
Here is the generated function signature.

```rust
 pub async fn get_one<P>(
            __rocket: &::rocket_sync_db_pools::rocket::Rocket<P>,
        ) -> Option<Self>
        where
            P: ::rocket_sync_db_pools::rocket::Phase,
```
The problem would be spotted immediately when these type requirements are added.

**Reply by ELD** [reply]
That's a fair question. I don't know.
